Add `vk::offset` to specify member offsets for push constants

fairywreath opened the pull request:
Closes #6660.

Adds a `vk::offset` modifier to specify offsets for struct fields. The main use-case for this is for push constants to specify member variables for different shader stages. The `vk::offset` is derived from DXC, this has already been implemented there.

jkwak-work commented on source/slang/slang-emit-glsl.cpp:
```diff
@@ -234,6 +234,26 @@ void GLSLSourceEmitter::emitMemoryQualifiers(IRInst* varDecl)
+void GLSLSourceEmitter::emitStructFieldAttributes(
+    IRStructType* structType,
+    IRStructField* field,
+    bool allowOffsetLayout)
```
Wouldn't it make more sense to take `IRGLSLStructOffsetDecoration` as the 3rd parameter?

fairywreath replied:
The problem is this function is not called directly but is a virtual function of the base c-like emitter class and is invoked only by this base class. Passing in `IRGLSLStructOffsetDecoration` directly would mean having the caller(that is for all C-like targets) perform a `findDecoration` instead of performing the find only on the GLSL emitter implementation. The `bool allowOffsetLayout` parameter is also consistent with this function's parent virtual function `emitStructDeclarationsBlock` that has the same exact bool parameter.

jkwak-work replied:
That makes sense.
Thanks for the explanation.

jkwak-work reviewed (commented):
It looks good to me.
But I am not sure if there is any reason why we may not want to support this new syntax.

fairywreath commented:
I renamed cases of `GLSLStructOffset` to `VkStructOffset` as this is mainly for Vulkan(regular struct member offsets are not supported in GLSL) and added an emit-spirv-via-glsl test cases in the latest commit.

jkwak-work reviewed (approved):
Looks good to me.
